### PR TITLE
S3 credentials for default user via existing secret

### PR DESCRIPTION
**Pull request by giubacc**
- Configuration options: `useExistingSecret` and `defaultUserCredentialsSecret`.
  These fields allow the user to specify an existing secret containing
  the S3 credentials for the default user.
- `useExistingSecret` is a boolean field defaulted to false.
- `defaultUserCredentialsSecret` is a string field denoting a `secret` in the `s3gw`
  namespace. It must contain 2 keys:
  - `RGW_DEFAULT_USER_ACCESS_KEY` that is the S3 Access Key for the default user.
  - `RGW_DEFAULT_USER_SECRET_KEY` that is the S3 Secret Key for the default user.
- When `useExistingSecret` is set to `false`, the chart will create
  the secret using values from the preexisting fields `accessKey` and `secretKey`.
- Setting `accessKey` or `secretKey` as the empty string, will force the Chart
  to compute random alphanumeric values for the fields.
- Defaulted values:
  - `useExistingSecret`: false
  - `defaultUserCredentialsSecret`: s3gw-creds

Tip: you can create a preexisting secret with a command like this:
 
```
kubectl create secret generic -n s3gw s3gw-creds --from-literal=accessKey=myaccesskey --from-literal=secretKey=supersecret
```

Fixes: https://github.com/aquarist-labs/s3gw/issues/303
Signed-off-by: Giuseppe Baccini <giuseppe.baccini@suse.com>

## Checklist before requesting a review

- [x] I have performed a self-review of my code.
- [x] If it is a core feature, I have added thorough tests.
- [x] CHANGELOG.md has been updated should there be relevant changes in this PR.

### CODE REVIEW

File: charts/s3gw/templates/secret.yaml
Comment (giubacc) [reply charts/s3gw/templates/secret.yaml:5]: are you suggesting to give the user the option to specify the names of the fields for the existing secret?

Comment (m-ildefons) [reply charts/s3gw/templates/secret.yaml:5]: If it is very obvious in the documentation that the secret must be set up this way, then it's ok.
I don't think namespaces are much of an issue, but if they are, we can still replace `.Release.Namespace` with a value from a user defined field, later on.

One thing that comes to my mind here is that these values are just used for initialization, right? So if a user supplies their custom secret, then deploys the s3gw, and then changes that custom secret, the s3gw won't pick up the change. This isn't a huge issue IMO, but should be mentioned in the documentation.
One way around that would be to require the user to supply `RGW_DEFAULT_USER_ACCESS_KEY` and `RGW_DEFAULT_USER_SECRET_KEY` as keys in their secret and then mount the user-supplied secret directly.

Comment (giubacc) [reply charts/s3gw/templates/secret.yaml:5]: > If it is very obvious in the documentation that the secret must be set up this way, then it's ok. 

ok, let's assume this will be very clear on doc.

>I don't think namespaces are much of an issue, but if they are, we can still replace `.Release.Namespace` with a value from a user defined field, later on.

ok
 
> One thing that comes to my mind here is that these values are just used for initialization, right? 

Yes, this secret is used on initialization phase; it is convenient when you are installing s3gw as subchart in another project.
This way, the embedder chart can also read from that secret and initialize all the S3 clients that will use s3gw.

> So if a user supplies their custom secret, then deploys the s3gw, and then changes that custom secret, the s3gw won't pick up the change. This isn't a huge issue IMO, but should be mentioned in the documentation. One way around that would be to require the user to supply `RGW_DEFAULT_USER_ACCESS_KEY` and `RGW_DEFAULT_USER_SECRET_KEY` as keys in their secret and then mount the user-supplied secret directly.

let's keep in mind that this is intended for initialization only.
`RGW_DEFAULT_USER_ACCESS_KEY` and `RGW_DEFAULT_USER_SECRET_KEY` are read by radosgw, but they are processed only if there is not already a default user in the system.
So, in the end, modifying those values won't change the default user credentials.
Moreover, this won't solve the problem that you still will have to change all the clients's configuration.

Comment (m-ildefons) [reply charts/s3gw/templates/secret.yaml:5]: In that case we don't lose any functionality, because we're using `RGW_DEFAULT_USER_ACCESS_KEY` and `RGW_DEFAULT_USER_SECRET_KEY` right now ourselves.
It should be made clear in the docs that these settings are used during initialization only. And I mean both our existing settings as well as this change.

But thanks for clarifying this.

**Review by m-ildefons — DISMISSED** [review]
Good to go after clarifying questions

**Comment by giubacc** [comment]
There is an unexpected behavior when trying to use s3gw as subchart with this patch.
Please, do not merge.

**Comment by jhmarina** [comment]
@giubacc is this something that needs to be documented? if so, I suggest that we update the documentation too.

**Comment by giubacc** [comment]
> @giubacc is this something that needs to be documented? if so, I suggest that we update the documentation too.

sure, doc needs to be updated

**Pull request merged**
